oxide-vpc/tests/integration_tests.rs

```diff
@@ -2218,204 +2217,356 @@ fn gateway_router_advert_reply() {
+    // Check that the inner IPs are what we expect.
+    let ip6 = if let IpMeta::Ip6(ip6) =
+        meta.inner.ip.as_ref().expect("No inner IP header")
+    {
+        ip6
+    } else {
+        panic!("Inner IP header is not IPv6");
     };
+    assert_eq!(ip6.src, na.src_ip);
+    assert_eq!(ip6.dst, na.dst_ip);
+    assert_eq!(ip6.proto, Protocol::ICMPv6);
 
+    // Validate the details of the Neighbor Advertisement itself.
     let rdr = reply.get_body_rdr();
     let reply_body = rdr.copy_remaining();
     let reply_pkt = Icmpv6Packet::new_checked(&reply_body).unwrap();
     let mut csum = CsumCapab::ignored();
     csum.icmpv6 = smoltcp::phy::Checksum::Rx;
-    let reply_icmp =
-        Icmpv6Repr::parse(&src.into(), &dst.into(), &reply_pkt, &csum).unwrap();
-    match reply_icmp {
-        Icmpv6Repr::Ndisc(NdiscRepr::NeighborAdvert {
-            flags,
-            target_addr,
+    let reply_icmp = Icmpv6Repr::parse(
+        &IpAddress::Ipv6(ip6.src.into()),
+        &IpAddress::Ipv6(ip6.dst.into()),
+        &reply_pkt,
+        &csum,
+    )
+    .unwrap();
+    if let Icmpv6Repr::Ndisc(NdiscRepr::NeighborAdvert {
+        flags,
+        target_addr,
+        lladdr,
+    }) = reply_icmp
+    {
+        assert_eq!(flags, na.flags);
+        assert_eq!(target_addr, na.target_addr.into());
+        assert_eq!(
             lladdr,
-        }) => {
-            assert_eq!(
-                flags,
-                NdiscNeighborFlags::ROUTER
-                    | NdiscNeighborFlags::SOLICITED
-                    | NdiscNeighborFlags::OVERRIDE
-            );
-            assert_eq!(
-                target_addr,
-                Ipv6Addr::from_eui64(&g1_cfg.gateway_mac).into()
-            );
-            assert_eq!(
-                lladdr.expect("Expected a Link-Layer Address option"),
-                RawHardwareAddress::from_bytes(&g1_cfg.gateway_mac.bytes())
-            );
-        }
-        other => {
-            panic!(
-                "Expected an ICMPv6 Neighbor Advertisement, found {:?}",
-                other
-            );
-        }
-    };
+            na.lladdr.map(|x| RawHardwareAddress::from_bytes(&x.bytes()))
+        );
+    } else {
+        panic!(
+            "Expected an ICMPv6 Neighbor Advertisement, found {:?}",
+            reply_icmp
+        );
+    }
+}
+
+// Ensure that we either Drop a Neighbor Solicitation, or generate a Neighbor
+// Advertisement with the right data, based on our defined test cases.
+#[test]
+fn test_gateway_neighbor_advert_reply() {
+    let g1_cfg = g1_cfg();
+    let v2p = Arc::new(Virt2Phys::new());
+    let mut ameta = ActionMeta::new();
+    let mut g1 = oxide_net_setup("g1_port", &g1_cfg, v2p.clone());
+    g1.port.start();
+    set_state!(g1, PortState::Running);
+    let mut pcap = PcapBuilder::new("gateway_neighbor_advert_reply.pcap");
+
+    let data = generate_solicit_test_data(&g1_cfg);
+    for d in data.into_iter().skip(1).take(1) {
```

Comment:
This looks like leftover from when you were trying to just test one case.

Reply:
Oh, now _that's_ a good catch, thank you!